app/models/account/gateway.rb

```diff
@@ -9,17 +9,15 @@ module Account::Gateway
     after_commit :update_payment_gateway, :if => :payment_gateway_changed
 
     attr_accessor :payment_gateway_changed
-
-
   end
 
   # Payment gateway this account accepts payments through.
   def payment_gateway
-    @payment_gateway ||=
-      if payment_gateway_type.present?
-        gateway_class = PaymentGateway.implementation(payment_gateway_type)
-        gateway_class.new(payment_gateway_options || {})
-      end
+    @payment_gateway ||= build_payment_gateway
+  end
+
+  def payment_gateway_setting_exists?
+    payment_gateway_configured? || events.payment_setting_deleted.any?
```

Comment:
I do not get this part. How existence of payment gateway setting is related to payment setting deleted

Imagine. I have a payment setting but I decided to delete it (for some reason)
Then month or years later my account is deleted ...

Reply:
This is for the error that @macejmic found when deleting in parallel:
```
  Exception -- {:exception=>{:class=>ArgumentError, :message=>"Missing required parameter: login", :backtrace=>["/home/bender/system_preview/shared/bundle/ruby/2.3.0/gems/activemerchant-1.77.0/lib/active_merchant/billing/gateway.rb:321:in `block in requires!'", "/home/bender/system_preview/shared/bundle/ruby/2.3.0/gems/activemerchant-1.77.0/lib/active_merchant/billing/gateway.rb:314:in `each'", "/home/bender/system_preview/shared/bundle/ruby/2.3.0/gems/activemerchant-1.77.0/lib/active_merchant/billing/gateway.rb:314:in `requires!'", "/home/bender/system_preview/shared/bundle/ruby/2.3.0/gems/activemerchant-1.77.0/lib/active_merchant/billing/gateways/stripe.rb:71:in `initialize'"]}, :parameters=>{:context=>"Job raised exception", :job=>{"class"=>"ActiveJob::QueueAdapters::SidekiqAdapter::JobWrapper", "wrapped"=>"DeletePlainObjectWorker", "queue"=>"default", "args"=>[{"job_class"=>"DeletePlainObjectWorker", "job_id"=>"bdcebe15-2c48-41a2-9d32-e9cace8f17c3", "queue_name"=>"default", "arguments"=>[{"_aj_globalid"=>"gid://system/Account/OBFUSCATEDID"}, ["Hierarchy-Account-OBFUSCATEDID"]], "locale"=>"en"}], "retry"=>true, "jid"=>"f604cefd953a3e991e79186f", "created_at"=>OBFUSCATEDDATE, "enqueued_at"=>1558367154.9197505, "error_message"=>"Missing required parameter: login", "error_class"=>"ArgumentError", "failed_at"=>1558366916.1332533, "retry_count"=>3, "retried_at"=>1558367155.9997318}, :jobstr=>"{\"class\":\"ActiveJob::QueueAdapters::SidekiqAdapter::JobWrapper\",\"wrapped\":\"DeletePlainObjectWorker\",\"queue\":\"default\",\"args\":[{\"job_class\":\"DeletePlainObjectWorker\",\"job_id\":\"bdcebe15-2c48-41a2-9d32-e9cace8f17c3\",\"queue_name\":\"default\",\"arguments\":[{\"_aj_globalid\":\"gid://system/Account/OBFUSCATEDID\"},[\"Hierarchy-Account-OBFUSCATEDID\"]],\"locale\":\"en\"}],\"retry\":true,\"jid\":\"f604cefd953a3e991e79186f\",\"created_at\":1558366914.7207932,\"enqueued_at\":1558367154.9197505,\"error_message\":\"Missing required parameter: login\",\"error_class\":\"ArgumentError\",\"failed_at\":1558366916.1332533,\"retry_count\":2,\"retried_at\":1558367041.2749653}"}}
```

Explained by @macejmic when he found out why:

> i finally got the problem with `ArgumentError: Missing required parameter: login`
> account has after destroy commit `before_destroy :unstore_credit_card!`
> but during that, settings are already gone..
> since the account is the last one that is being deleted.. there are not payment/gateway settings..
> so the credit card could not be unstored..

So what this code does is, at the moment of the deletion of the account: do you have a payment gateway config persisted or an event that means that has been deleted right now? if so, make the unstore, otherwise, do nothing related to unstoring credit card because we can't do miracle with info that doesn't exist anywhere anymore 😄 

So in your example:

> Imagine. I have a payment setting but I decided to delete it (for some reason)
> Then month or years later my account is deleted ...

Then it would completely ignore at this point the fact that the payment gateway was deleted so long ago 😄 but it wouldn't raise an error 😄

Reply:
The looks like a hack.